internal/report/report.go

```diff
@@ -0,0 +1,218 @@
+			if err != nil {
+				return []vulnerability{}, fmt.Errorf("is vulnerability excluded: %w", err)
+			}
+			v := vulnerability{
+				CheckData:     check.CheckData,
+				Vulnerability: vuln,
+				Severity:      severity,
+				excluded:      excluded,
+			}
+			vulns = append(vulns, v)
+		}
+	}
+	return vulns, nil
+}
+
+// vulnerability represents each vulnerability found by a check.
+type vulnerability struct {
+	CheckData            report.CheckData `json:"check_data"`
+	report.Vulnerability `json:"vulnerability"`
+	Severity             config.Severity `json:"severity"`
+	excluded             bool
+}
+
+// filterVulns takes a list of vulnerabilities and
+// filters them out ready to be displayed in the final report.
+func filterVulns(vulns []vulnerability, severity config.Severity) []vulnerability {
+	// Sort the results by severity in reverse order.
+	slices.SortFunc(vulns, func(a, b vulnerability) int {
+		return cmp.Compare(b.Severity, a.Severity)
+	})
+
+	var vulnerabilities []vulnerability
+	for _, v := range vulns {
+		if v.Severity < severity {
+			break
+		}
+		if v.excluded {
+			continue
+		}
+		vulnerabilities = append(vulnerabilities, v)
+	}
+	return vulnerabilities
+}
+
+// printer is the interface to represent every kind of report writer.
+type printer interface {
+	print(vulns []vulnerability, stats stats) error
+}
+
+// scoreToSeverity calculates the severity using the score.
+// To calculate the severity we are using the [severity ratings]
+// provided by the NVD.
+//
+// [severity ratings]: https://nvd.nist.gov/vuln-metrics/cvss
+func scoreToSeverity(score float32) config.Severity {
+	switch {
+	case score >= 9.0:
+		return config.SeverityCritical
+	case score >= 7.0:
+		return config.SeverityHigh
+	case score >= 4.0:
+		return config.SeverityMedium
+	case score >= 0.1:
+		return config.SeverityLow
+	default:
+		return config.SeverityInfo
+	}
+}
+
+// isExcluded returns true if any of the exclusions matches with the vulnerability.
+func isExcluded(v report.Vulnerability, target string, excls []config.Exclusion) (bool, error) {
+	for _, excl := range excls {
+		summaryRegexp, err := regexp.Compile(excl.Summary)
+		if err != nil {
+			return false, fmt.Errorf("compile summary regex: %w", err)
+		}
+
+		if strings.Contains(target, excl.Target) &&
+			summaryRegexp.MatchString(v.Summary) &&
+			strings.Contains(v.Fingerprint, excl.Fingerprint) &&
+			(strings.Contains(v.AffectedResource, excl.Resource) || strings.Contains(v.AffectedResourceString, excl.Resource)) {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// stats represents the statistics of the results.
+type stats struct {
+	count    map[config.Severity]int
+	excluded int
+}
+
+// calculateStats counts how many vulnerabilities of each type there are and
+// how many vulnerabilities are excluded. The excluded vulnerabilities are not
+// included in the total amount of each severity.
+func calculateStats(vulns []vulnerability) (stats, error) {
```

Comment:
I would rename this function to something like:

```suggestion
func mkSummary(vulns []vulnerability) (summary, error) {
```